### PR TITLE
Expand directories in Go srcs param

### PR DESCRIPTION
**Pull request by linzhp**
As discussed in #518 and #560, to compile the output files of a genrule, we need to make a genrule for each output. This could be problematic when the first genrule produces multiple files and their names are not predictable.

This PR makes it possible to put such genrule target in the `srcs` and compile the all generated files.

### CODE REVIEW

**Comment by linzhp** [comment]
@styurin could you take a look?

File: src/com/facebook/buck/go/GoCompile.java [hunk 1]
```diff
@@ -108,15 +113,27 @@ public GoCompile(
     ImmutableList.Builder<Path> compileSrcListBuilder = ImmutableList.builder();
     ImmutableList.Builder<Path> headerSrcListBuilder = ImmutableList.builder();
     ImmutableList.Builder<Path> asmSrcListBuilder = ImmutableList.builder();
+    List<Path> srcFiles = new ArrayList<>();
```
Comment (styurin) [inline src/com/facebook/buck/go/GoCompile.java]: This method is already too big, could you move the logic responsible for collecting source files into a standalone method?

File: src/com/facebook/buck/go/GoCompile.java [hunk 2]
```diff
@@ -108,15 +113,27 @@ public GoCompile(
...
+        try {
+          srcFiles.addAll(Files.list(srcPath).filter(Files::isRegularFile).collect(Collectors.toList()));
+        } catch (IOException e) {
+          throw new RuntimeException("An error occur when listing the files under " + srcPath, e);
+        }
+      } else {
+        srcFiles.add(srcPath);
+      }
+    }
+    for (Path f : srcFiles) {
```
Comment (styurin) [inline src/com/facebook/buck/go/GoCompile.java]: Please, use full work, `f` is not a well-known reduction

**Comment by linzhp** [comment]
@styurin All requested changes are made. Could you take another look?

File: src/com/facebook/buck/go/GoCompile.java [hunk 3]
```diff
@@ -255,6 +260,23 @@ public GoCompile(
     return steps.build();
   }
 
+  private List<Path> getSourchFiles(BuildContext context) {
```
Comment (styurin) [inline src/com/facebook/buck/go/GoCompile.java]: Should be `getSourceFiles`

**Comment by styurin** [comment]
Could reformat source using google-java-format? Here are the instructions: https://github.com/facebook/buck/blob/master/CONTRIBUTING.md#code-style

**Comment by linzhp** [comment]
@styurin Let's move the discussion on the directory support here:

We've considered the zipping option before implementing this PR, but felt it might slow down the build. And I couldn't remember if I figured out how to unzip the source files in go_library either. We are open to more discussion on that issue.

**Comment by styurin** [comment]
You can use `UnzipStep` to unzip an archive.

**Comment by linzhp** [comment]
The output of UnzipStep is a directory. So in the end, we still have to deal with source directories anyway. Zipping and unzipping merely add some unnecessary computational cost. Do you see any benefits?

**Pull request closed without merging**